Added Non-Generic Solution to KiotaDeserialization

DerGuru opened the pull request:
This code addition fulfills the request of [Issue 210](https://github.com/microsoft/kiota-dotnet/issues/210), where this was explicitly requested.

@dansmitt is a co-worker of mine and I thought, that this might be of relevance to more than just him.

I used the same Unit-Tests, as you were, modifying them just as much, to use the newly introduced method signatures.

baywet commented on src/abstractions/serialization/KiotaJsonSerializer.Deserialization.Untyped.cs:
```diff
@@ -0,0 +1,88 @@
+public static partial class KiotaJsonSerializer
+{
+    private static bool IsIParsable(this Type? type) => type?.IsAssignableTo(typeof(IParsable)) ?? false;
+    private abstract class KiotaJsonDeserializer
```
can we give different names to this class and the one bellow to ease up code review?
It feels like this one should be called somethingFactory.
And the one at line 30 somethingAdapter.

Or something alone those lines.

DerGuru replied:
Hi baywet,

surely you could name them as you please, but to comment on your suggestions.
The factory is in fact an abstract base class having a factory method (well one line of it) and executes the actual Activator, which is then placed in line 30.

I created this much like in System.Security, where (f. e.) one calls the create method of the abstract AES, which returns an actual implementation of a derived class.

I tried to keep it as separate as possible and even the derived classes are privately nested classes of the abstract base.

If you want to call the class something different, you might call it a wrapper around the KitoJsonSerializer for this special case of having the target type as Type instead of handing it to a generic method. So I guess I will rename it.

I also got e broken build, which was the result of some wrong line endings and different whitespacing.
Is there something I could download for VS to have the style guide enforced before pushing it again to GitHub?

baywet commented on src/abstractions/serialization/KiotaJsonSerializer.Deserialization.Untyped.cs:
```diff
@@ -0,0 +1,88 @@
+using System.Collections.Generic;
+using System.Linq;
+
+#pragma warning disable IL3050 
```
can you provide more context around this suppression please?

DerGuru replied:
Well the Warning I received was, that AOT will not have the type ready when using a native assembly and will have to create it during runtime. At least that is, what I got from the message and some reading on the internet.
 Nevertheless, this is intended behavior, because this is exactly for the case, when one does not know the Type at compile time and therefore needs this API in the first place. Otherwise one could have used the Generic Methods instead of these.

baywet reviewed (changes requested):
Thanks for the contribution!

DerGuru commented:
Okay .. @baywet
I took your comments to heart and split the base class into a factory and an interface.
I also hope to have fixed the whitespace issue.... weirdly, that was auto generated code...

I was also able, to remove the `#pragma warning disable IL3050`, which just needed some attributes added to the code. And since I don't like to have pre compiler directives spread all over the code, I added the Attribute as a privately nested class, for .NET < 8. 
I hope, that is okay.

baywet commented on src/abstractions/serialization/KiotaJsonSerializer.Deserialization.NonGeneric.cs:
```diff
@@ -0,0 +1,105 @@
+    /// <summary>
+    /// Palceholder for .NET &lt; 8
+    /// </summary>
+    private class RequiresDynamicCodeAttribute : Attribute
```
Here is how we've solved the problem across the board.
https://github.com/microsoft/kiota-dotnet/blob/650812d9b598f192e6e55424e327d2bcee49c99a/src/abstractions/RequestInformation.cs#L51
Can you align with this please?

DerGuru replied:
done

DerGuru replied:
Well it looks kind of ugly now ... but can't help that, I guess.

baywet replied:
yes it does. That's ok, consistency over "prettyness". Once we address this, we'll be able to do so across the libraries.

baywet commented on src/abstractions/serialization/KiotaJsonSerializer.Deserialization.NonGeneric.cs:
```diff
@@ -0,0 +1,105 @@
+    /// <summary>
+    /// Deserializes the given string into an object.
+    /// </summary>
+    /// <param name="targetType">The target type to deserialize</param>
+    /// <param name="serializedRepresentation">The serialized representation of the object.</param>
+    /// <param name="cancellationToken">The cancellation token for the task</param>
+    [RequiresDynamicCode("Activator creates an instance of a generic class with the Target Type as the generic type argument.")]
+    public static Task<IParsable?> DeserializeAsync([DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods)] Type targetType, string serializedRepresentation, CancellationToken cancellationToken = default)
+        => KiotaJsonDeserializationWrapperFactory.Create(targetType).DeserializeAsync(serializedRepresentation, cancellationToken);
+
+    /// <summary>
+    /// Deserializes the given stream into an object.
+    /// </summary>
+    /// <param name="targetType">The target type to deserialize</param>
+    /// <param name="stream">The stream to deserialize.</param>
+    /// <param name="cancellationToken">The cancellation token for the task</param>
+    [RequiresDynamicCode("Activator creates an instance of a generic class with the Target Type as the generic type argument.")]
+    public static Task<IParsable?> DeserializeAsync([DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods)] Type targetType, Stream stream, CancellationToken cancellationToken = default)
+         => KiotaJsonDeserializationWrapperFactory.Create(targetType).DeserializeAsync(stream, cancellationToken);
+
+    /// <summary>
+    /// Deserializes the given stream into a collection of objects based on the content type.
+    /// </summary>
+    /// <param name="targetType">The target type to deserialize</param>
+    /// <param name="stream">The stream to deserialize.</param>
+    /// <param name="cancellationToken">The cancellation token for the task</param>
+    [RequiresDynamicCode("Activator creates an instance of a generic class with the Target Type as the generic type argument.")]
+    public static Task<IEnumerable<IParsable>> DeserializeCollectionAsync(Type targetType, Stream stream, CancellationToken cancellationToken = default)
+         => KiotaJsonDeserializationWrapperFactory.Create(targetType).DeserializeCollectionAsync(stream, cancellationToken);
+
+    /// <summary>
+    /// Deserializes the given stream into a collection of objects based on the content type.
+    /// </summary>
+    /// <param name="targetType">The target type to deserialize</param>
+    /// <param name="serializedRepresentation">The serialized representation of the object.</param>
+    /// <param name="cancellationToken">The cancellation token for the task</param>
+    [RequiresDynamicCode("Activator creates an instance of a generic class with the Target Type as the generic type argument.")]
+    public static Task<IEnumerable<IParsable>> DeserializeCollectionAsync(Type targetType, string serializedRepresentation, CancellationToken cancellationToken = default)
+         => KiotaJsonDeserializationWrapperFactory.Create(targetType).DeserializeCollectionAsync(serializedRepresentation, cancellationToken);
+}
```
why is this whole section duplicated instead of calling into the non media type specific one?

baywet reviewed (commented):
Thank you for making the changes!
A couple additional questions/comments.

baywet reviewed (dismissed):
Thank you for making the changes!

I also took the liberty of making a bunch of "cleanup" changes here.

@andrueastman for final review and merge

andrueastman reviewed (commented):
Otherwise looks good from my end. Will update version/release notes on getting clarification on duplicate code question.
